src/python/pants/backend/docker/rules.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+import os
+from dataclasses import dataclass
+
+from pants.backend.docker.rules_binary import DockerBinary, DockerBinaryRequest
+from pants.backend.docker.rules_binary import rules as binary_rules
+from pants.backend.docker.rules_context import DockerBuildContext, DockerBuildContextRequest
+from pants.backend.docker.rules_context import rules as context_rules
+from pants.backend.docker.target_types import (
+    DockerContextRoot,
+    DockerImageSources,
+    DockerImageVersion,
+)
+from pants.core.goals.package import BuiltPackage, BuiltPackageArtifact, PackageFieldSet
+from pants.engine.fs import Digest, Snapshot
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import TransitiveTargets, TransitiveTargetsRequest
+from pants.engine.unions import UnionRule
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class DockerFieldSet(PackageFieldSet):
+    required_fields = (DockerImageSources,)
+
+    context_root_field: DockerContextRoot
+    image_version: DockerImageVersion
+    sources: DockerImageSources
+
+    @property
+    def dockerfile_name(self) -> str:
+        if not self.sources.value:
```

Comment:
I don't think this case is possible, since DockerImageSources has `expected_num_files = 1`, which is validated.  And the default of `Dockerfile` is already set on that class and we don't want to repeat it here. So I think we can either not validate here at all, or fail hard here if this validation fails, since it's not supposed to be possible. @Eric-Arellano , am I right about that?

Reply:
Ah, this could be the reason why I added that check:
```
src/python/pants/backend/docker/docker_build.py:35:16: error: Value of type
"Optional[Tuple[str, ...]]" is not indexable  [index]
            return self.sources.value[0]
                   ^
Found 1 error in 1 file (checked 204 source files)
```

Using a `cast` instead, is that the proper way to deal with it?